RecoMuon/TrackerSeedGenerator/src/SeedMvaEstimator.cc

```diff
@@ -0,0 +1,132 @@
+#include "DataFormats/Math/interface/deltaR.h"
+
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+#include "CommonTools/MVAUtils/interface/GBRForestTools.h"
+
+SeedMvaEstimator::SeedMvaEstimator(const edm::FileInPath& weightsfile,
+                                   std::vector<double> scale_mean,
+                                   std::vector<double> scale_std) {
+  gbrForest_ = createGBRForest(weightsfile);
+  scale_mean_ = scale_mean;
+  scale_std_ = scale_std;
+}
+
+SeedMvaEstimator::~SeedMvaEstimator() {}
+
+namespace {
+  enum inputIndexes {
+    kTsosErr0,       // 0
+    kTsosErr2,       // 1
+    kTsosErr5,       // 2
+    kTsosDxdz,       // 3
+    kTsosDydz,       // 4
+    kTsosQbp,        // 5
+    kDRdRL1SeedP,    // 6
+    kDPhidRL1SeedP,  // 7
+    kLastL1,         // 8
+
+    kDRdRL2SeedP = 8,  // 8
+    kDPhidRL2SeedP,    // 9
+    kLastL2,           // 10
+  };
+}  // namespace
+
+void SeedMvaEstimator::getL1MuonVariables(const TrajectorySeed& seed,
+                                          const GlobalVector& global_p,
+                                          const l1t::MuonBxCollection& L1Muons,
+                                          int minL1Qual,
+                                          float& dR2dRL1SeedP,
+                                          float& dPhidRL1SeedP) const {
+  for (int ibx = L1Muons.getFirstBX(); ibx <= L1Muons.getLastBX(); ++ibx) {
+    if (ibx != 0)
+      continue;  // -- only take when ibx == 0 -- //
+
+    for (auto it = L1Muons.begin(ibx); it != L1Muons.end(ibx); it++) {
+      if (it->hwQual() < minL1Qual)
+        continue;
+
+      float dR2tmp = reco::deltaR2(it->etaAtVtx(), it->phiAtVtx(), global_p.eta(), global_p.phi());
+      if (dR2tmp < dR2dRL1SeedP) {
+        dR2dRL1SeedP = dR2tmp;
+        dPhidRL1SeedP = reco::deltaPhi(it->phiAtVtx(), global_p.phi());
+      }
+    }
+  }
+}
+
+void SeedMvaEstimator::getL2MuonVariables(const TrajectorySeed& seed,
+                                          const GlobalVector& global_p,
+                                          const reco::RecoChargedCandidateCollection& L2Muons,
+                                          float& dR2dRL2SeedP,
+                                          float& dPhidRL2SeedP) const {
+  for (auto it = L2Muons.begin(); it != L2Muons.end(); it++) {
+    float dR2tmp = reco::deltaR2(*it, global_p);
+    if (dR2tmp < dR2dRL2SeedP) {
+      dR2dRL2SeedP = dR2tmp;
+      dPhidRL2SeedP = reco::deltaPhi(it->phi(), global_p.phi());
+    }
+  }
+}
+
+double SeedMvaEstimator::computeMva(const TrajectorySeed& seed,
+                                    const GlobalVector& global_p,
+                                    const l1t::MuonBxCollection& L1Muons,
+                                    int minL1Qual,
+                                    const reco::RecoChargedCandidateCollection& L2Muons,
+                                    bool isFromL1) const {
+  float initDRdPhi = 99999.;
```

Comment:
Thanks for the comments again. You taught me a lot.
When I compile or run this code, this works well. However when I do (scram build code-checks), float var[kLast]{} makes error (variable-sized object may not be initialized [clang-diagnostic-error]) even though kLast is const.
float var[(isFromL1? kLastL1 : kLastL2)]{} makes error too.
```
float varL1[kLastL1]{};
float varL2[kLastL2]{};
float* var = isFromL1 ? varL1 : varL2;
```
This seems working well, but I am not sure this way is efficient. Could you give me any advice?

Reply:
Good point: `code-cheks` does not allow the size of the C-style vectors being defined at run time.

I got something working and code-checks compliant using `union`:

```C++
  static constexpr float initDRdPhi(99999.);
  const int kLast = isFromL1 ? kLastL1 : kLastL2;

  union theVars {float var[kLastL2]; float varL1[kLastL1]; float varL2[kLastL2];};
  theVars var{};

  var.var[kTsosErr0] = seed.startingState().error(0);
  var.var[kTsosErr2] = seed.startingState().error(2);
  var.var[kTsosErr5] = seed.startingState().error(5);
  var.var[kTsosDxdz] = seed.startingState().parameters().dxdz();
  var.var[kTsosDydz] = seed.startingState().parameters().dydz();
  var.var[kTsosQbp] = seed.startingState().parameters().qbp();

  float dR2dRL1SeedP = initDRdPhi;
  float dPhidRL1SeedP = initDRdPhi;
  getL1MuonVariables(seed, global_p, L1Muons, minL1Qual, dR2dRL1SeedP, dPhidRL1SeedP);

  var.var[kDRdRL1SeedP] = std::sqrt(dR2dRL1SeedP);
  var.var[kDPhidRL1SeedP] = dPhidRL1SeedP;

  if (!isFromL1) {
    float dR2dRL2SeedP = initDRdPhi;
    float dPhidRL2SeedP = initDRdPhi;
    getL2MuonVariables(seed, global_p, L2Muons, dR2dRL2SeedP, dPhidRL2SeedP);

    var.var[kDRdRL2SeedP] = std::sqrt(dR2dRL2SeedP);
    var.var[kDPhidRL2SeedP] = dPhidRL2SeedP;
  }

  for (int iv = 0; iv < kLast; ++iv) {
    var.var[iv] = (var.var[iv] - scale_mean_.at(iv)) / scale_std_.at(iv);
  }

  if (isFromL1)
    return gbrForest_->GetResponse(var.varL1);
  else
    return gbrForest_->GetResponse(var.varL2);
```

However, before proceeding I would like to check with the C++ gurus (e.g. @cms-sw/core-l2 , @slava77 ...) whether this is a legitimate usage of a `union`. According to [this link](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#SS-union) I would say "yes", because it is not used here to cast a different type, but simply to consider arrays of different lengths,

Moreover, one should also verify that the correct vectors are propagated to the gbrForest: I only verified that this code builds, and satisfies the code-checks

Reply:
> Good point: `code-cheks` does not allow the size of the C-style vectors being defined at run time.

It's not just that, such variable length arrays are not part of C++ standard.

Strictly speaking this use of `union` is undefined behavior. The types in the `union` are different, and writing to one and reading from other would goes under the "type punning"  guideline [C.183](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ru-pun).

It seems to me that the `union` could be easily avoided. The union would always allocate space for `kLastL2` elements anyway, so how about just allocating array of that size always along
```cpp
float var[kLastL2]{};
// fill var up to kLastL1
if (!isFromL1) {
  // fill var up to kLastL2
}
...
return gbrForest_->GetResponse(var);
```
?

How does `gbrForest_->GetResponse()` know which size array to expect as an argument?

Reply:
why are you initializing the  array? 
```float var[(isFromL1? kLastL1 : kLastL2)];```
should give NO error.

Reply:
and remove that union!

Reply:
> why are you initializing the array?
> `float var[(isFromL1? kLastL1 : kLastL2)];`
> should give NO error.

It builds, but it gives errors after `scram b code-checks`, see also https://github.com/cms-sw/cmssw/pull/33983#discussion_r651995273

Reply:
it says that you cannot initialized a run-time-sized array. so remove the initalization! ```{}```

Reply:
> loat varL1[kLastL1]{};
> float varL2[kLastL2]{};
> float* var = isFromL1 ? varL1 : varL2;
and btw this is perfectly efficient and better then the union!

Reply:
> it says that you cannot initialized a run-time-sized array. so remove the initalization! the curly brakets `{}`

Good point @VinInn !
Can you give it a try, @wonpoint4 ?

Reply:
Both of ```float var[kLastL2]{};``` and ```float var[(isFromL1? kLastL1 : kLastL2)];``` work well and give the same results.
Then I will delete the union, and use the latter way. Thanks for the tips!!

Reply:
I'm still curious (especially for the variable length array case) how `gbrForest_->GetResponse()` knows the number of elements in the array?

Reply:
```gbrForest_->GetResponse()``` uses this function (https://github.com/cms-sw/cmssw/blob/master/CondFormats/GBRForest/interface/GBRForest.h#L48-L54 )
And this uses ( https://github.com/cms-sw/cmssw/blob/master/CondFormats/GBRForest/interface/GBRTree.h#L62-L70 )

I guess the size of array dosen't matter if the size is larger than the number of BDT variables. Also when I test var[1000], it makes the same mva distributions. I thought it'd better to explicitly fix the number of variables before conversion into z values ( https://github.com/wonpoint4/cmssw/blob/MuonHLTSeedMVA_120X/RecoMuon/TrackerSeedGenerator/src/SeedMvaEstimator.cc#L112 )

Reply:
Thanks. Is `isFromL1` related to the model ("number of BDT variables") or something set e.g. event by event or object by object?

I'm thinking about what happens when it is passed the shorter array (of 8 elements). Can the BDT in any circumstance read the elements 8 and 9? Or, in other words, does the BDT somehow know which of the 8 or 10 element arrays is given to it, or does it always read 10 elements?

Reply:
> Thanks. Is `isFromL1` related to the model ("number of BDT variables") or something set e.g. event by event or object by object?
> 
> I'm thinking about what happens when it is passed the shorter array (of 8 elements). Can the BDT in any circumstance read the elements 8 and 9? Or, in other words, does the BDT somehow know which of the 8 or 10 element arrays is given to it, or does it always read 10 elements?

Indeed this is a (set of) good question(s).

As far as I can see from the GBRtree [1] the dimension of the vector is set by the weightfile that defines it. The ones set here by default for barrel and endcap muons are [both](https://github.com/cms-sw/cmssw/pull/33983/files#diff-6e54166e728761604c7df361175a2f0d93d9a22922360549afb4ed2725c543c6R258-R261) requesting the dimension 10 vector that corresponds to `isFromL1 = false`.

Said that, I see now some possible flaw with the current configuration:
- what happens if `isFromL1` is set at a value which is incompatible with the weightfiles defined at config?
- what happens if the weightfile for barrel is by mistake configured inconsistently with the one in the endcap?

A possible way out could be to allow in the configuration both pairs of files (e.g. "mvaFileBL1" "mvaFileEL1" and "mvaFileBL2" "mvaFileEL2"), and depending on the value of the configuration parameter  `isFromL1` select the two that correspond to the related configuration [2]. As such, one would switch everywhere between the dim-10 vector and dim-8 vector coherently throughout the whole code. Do you agree?

[1] https://cmssdt.cern.ch/lxr/source/CondFormats/GBRForest/interface/GBRTree.h#0062

[2] For example:
```C++
   mvaFileB_(iConfig.getParameter&lt;edm::FileInPath&gt;(isFromL1_? "mvaFileBL1":"mvaFileBL2")),      
   mvaFileE_(iConfig.getParameter&lt;edm::FileInPath&gt;(isFromL1_? "mvaFileEL1":"mvaFileEL2")),
```

Reply:
Thanks a lot for the comments.
As you noticed, if ```isFromL1``` is true, BTD model with 8 variables should be used, otherwise model with 10 variables should be used. We have trained 4 different models (Barrel or Endcap, and ```isFromL1``` true or false), and here ```isFromL1``` is related to the different tracking iterations inside L3 muon reconstruction sequence (seeded by only L1Muons or both of L1, L2Muons) 

Anyway I think perrotta's suggestion is very reasonable , and I'll give it a try. Thanks!